sql/tests: TestDescriptorRepairOrphanedDescriptors failed

[(sql/tests).TestDescriptorRepairOrphanedDescriptors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2695461&tab=buildLog) on [master@cb0d14a39c32772494a42362fbbcbc308dbf4b35](https://github.com/cockroachdb/cockroach/commits/cb0d14a39c32772494a42362fbbcbc308dbf4b35):

```
=== RUN   TestDescriptorRepairOrphanedDescriptors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors689702253
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestDescriptorRepairOrphanedDescriptors
    repair_test.go:246: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors689702253
--- FAIL: TestDescriptorRepairOrphanedDescriptors (11.07s)
=== RUN   TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    repair_test.go:146: 
        	Error Trace:	repair_test.go:146
        	Error:      	Received unexpected error:
        	            	pq: setting updated but timed out waiting to read new value
        	Test:       	TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    --- FAIL: TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782 (10.51s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDescriptorRepairOrphanedDescriptors PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDescriptorRepairOrphanedDescriptors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/tests).TestDescriptorRepairOrphanedDescriptors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2698187&tab=buildLog) on [master@8b6f3c84cc256debeeb4d4055c5f0d5c9a481213](https://github.com/cockroachdb/cockroach/commits/8b6f3c84cc256debeeb4d4055c5f0d5c9a481213):

```
=== RUN   TestDescriptorRepairOrphanedDescriptors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors705005801
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestDescriptorRepairOrphanedDescriptors
    repair_test.go:246: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors705005801
--- FAIL: TestDescriptorRepairOrphanedDescriptors (10.54s)
=== RUN   TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782
    repair_test.go:93: 
        	Error Trace:	repair_test.go:93
        	Error:      	Received unexpected error:
        	            	pq: setting updated but timed out waiting to read new value
        	Test:       	TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782
    --- FAIL: TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782 (10.26s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDescriptorRepairOrphanedDescriptors PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDescriptorRepairOrphanedDescriptors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/tests).TestDescriptorRepairOrphanedDescriptors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2702188&tab=buildLog) on [master@bf9744bad5a416a4b06907f0f3dd42896f7342f3](https://github.com/cockroachdb/cockroach/commits/bf9744bad5a416a4b06907f0f3dd42896f7342f3):

```
=== RUN   TestDescriptorRepairOrphanedDescriptors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors733677862
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestDescriptorRepairOrphanedDescriptors
    repair_test.go:246: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors733677862
--- FAIL: TestDescriptorRepairOrphanedDescriptors (11.19s)
=== RUN   TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782
    repair_test.go:93: 
        	Error Trace:	repair_test.go:93
        	Error:      	Received unexpected error:
        	            	pq: setting updated but timed out waiting to read new value
        	Test:       	TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782
    --- FAIL: TestDescriptorRepairOrphanedDescriptors/orphaned_view_-_51782 (10.54s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDescriptorRepairOrphanedDescriptors PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDescriptorRepairOrphanedDescriptors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/tests).TestDescriptorRepairOrphanedDescriptors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2703023&tab=buildLog) on [master@5b88d8000a062d01c0ea12f451d2b3a5e5bde432](https://github.com/cockroachdb/cockroach/commits/5b88d8000a062d01c0ea12f451d2b3a5e5bde432):

```
=== RUN   TestDescriptorRepairOrphanedDescriptors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors168340303
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestDescriptorRepairOrphanedDescriptors
    repair_test.go:246: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors168340303
--- FAIL: TestDescriptorRepairOrphanedDescriptors (10.41s)
=== RUN   TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    repair_test.go:146: 
        	Error Trace:	repair_test.go:146
        	Error:      	Received unexpected error:
        	            	pq: setting updated but timed out waiting to read new value
        	Test:       	TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    --- FAIL: TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782 (10.23s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDescriptorRepairOrphanedDescriptors PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDescriptorRepairOrphanedDescriptors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/tests).TestDescriptorRepairOrphanedDescriptors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2709471&tab=buildLog) on [master@bdcdf8810bbeb80c87d7a4f6f69cdbf460fa9940](https://github.com/cockroachdb/cockroach/commits/bdcdf8810bbeb80c87d7a4f6f69cdbf460fa9940):

```
=== RUN   TestDescriptorRepairOrphanedDescriptors
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors229937288
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestDescriptorRepairOrphanedDescriptors
    repair_test.go:246: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDescriptorRepairOrphanedDescriptors229937288
--- FAIL: TestDescriptorRepairOrphanedDescriptors (10.93s)
=== RUN   TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    repair_test.go:146: 
        	Error Trace:	repair_test.go:146
        	Error:      	Received unexpected error:
        	            	pq: setting updated but timed out waiting to read new value
        	Test:       	TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782
    --- FAIL: TestDescriptorRepairOrphanedDescriptors/orphaned_table_with_data_-_51782 (10.43s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDescriptorRepairOrphanedDescriptors PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDescriptorRepairOrphanedDescriptors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -1363,9 +1363,9 @@ func (tc *Collection) addUncommittedDescriptor(
 	return ud, nil
 }
 
-// validateOnWriteEnabled is the cluster setting used to enable or disable
+// ValidateOnWriteEnabled is the cluster setting used to enable or disable
 // validating descriptors prior to writing.
-var validateOnWriteEnabled = settings.RegisterBoolSetting(
+var ValidateOnWriteEnabled = settings.RegisterBoolSetting(
 	"sql.catalog.descs.validate_on_write.enabled",
 	"set to true to validate descriptors prior to writing, false to disable; default is true",
 	true, /* defaultValue */
@@ -1377,7 +1377,7 @@ func (tc *Collection) WriteDescToBatch(
 	ctx context.Context, kvTrace bool, desc catalog.MutableDescriptor, b *kv.Batch,
 ) error {
 	desc.MaybeIncrementVersion()
-	if validateOnWriteEnabled.Get(&tc.settings.SV) {
+	if ValidateOnWriteEnabled.Get(&tc.settings.SV) {
 		if err := desc.ValidateSelf(ctx); err != nil {
 			return err
 		}
@@ -1449,7 +1449,7 @@ func (cdg collectionDescGetter) GetDesc(
 
 // ValidateUncommittedDescriptors validates all uncommitted descriptors
 func (tc *Collection) ValidateUncommittedDescriptors(ctx context.Context, txn *kv.Txn) error {
-	if !validateOnWriteEnabled.Get(&tc.settings.SV) {
+	if !ValidateOnWriteEnabled.Get(&tc.settings.SV) {
 		return nil
 	}
 	cdg := collectionDescGetter{tc: tc, txn: txn}

--- a/pkg/sql/tests/BUILD.bazel
+++ b/pkg/sql/tests/BUILD.bazel
@@ -66,6 +66,7 @@ go_test(
         "//pkg/sql/catalog/bootstrap",
         "//pkg/sql/catalog/catalogkv",
         "//pkg/sql/catalog/descpb",
+        "//pkg/sql/catalog/descs",
         "//pkg/sql/catalog/systemschema",
         "//pkg/sql/parser",
         "//pkg/sql/pgwire/pgcode",

--- a/pkg/sql/tests/repair_test.go
+++ b/pkg/sql/tests/repair_test.go
@@ -19,6 +19,7 @@ import (
 
 	"github.com/cockroachdb/cockroach-go/crdb"
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -40,8 +41,6 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 	ctx := context.Background()
 	setup := func(t *testing.T) (serverutils.TestServerInterface, *gosql.DB, func()) {
 		s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
-		_, err := db.Exec(`SET CLUSTER SETTING sql.catalog.descs.validate_on_write.enabled = false`)
-		require.NoError(t, err)
 		return s, db, func() {
 			s.Stopper().Stop(ctx)
 		}
@@ -74,9 +73,10 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 	// descriptor and namespace entry. This would normally be unsafe because
 	// it would leave table data around.
 	t.Run("orphaned view - 51782", func(t *testing.T) {
-		_, db, cleanup := setup(t)
+		s, db, cleanup := setup(t)
 		defer cleanup()
 
+		descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, false)
 		require.NoError(t, crdb.ExecuteTx(ctx, db, nil, func(tx *gosql.Tx) error {
 			if _, err := tx.Exec(
 				"SELECT crdb_internal.unsafe_upsert_descriptor($1, decode($2, 'hex'));",
@@ -87,15 +87,12 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 				parentID, schemaID, tableName, descID)
 			return err
 		}))
-
-		// Now that we've finished setting up the test, we can restore validations.
-		_, err := db.Exec(`SET CLUSTER SETTING sql.catalog.descs.validate_on_write.enabled = true`)
-		require.NoError(t, err)
+		descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, true)
 
 		// Ideally we should be able to query `crdb_internal.invalid_object` but it
 		// does not do enough validation. Instead we'll just observe the issue that
 		// the parent descriptor cannot be found.
-		_, err = db.Exec(
+		_, err := db.Exec(
 			"SELECT count(*) FROM \"\".crdb_internal.tables WHERE table_id = $1",
 			descID)
 		require.Regexp(t, "internal error: desc 53: parentID 52 does not exist", err)
@@ -127,9 +124,10 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 	// will then repair it by injecting a new database descriptor and namespace
 	// entry and then demonstrate the problem is resolved.
 	t.Run("orphaned table with data - 51782", func(t *testing.T) {
-		_, db, cleanup := setup(t)
+		s, db, cleanup := setup(t)
 		defer cleanup()
 
+		descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, false)
 		require.NoError(t, crdb.ExecuteTx(ctx, db, nil, func(tx *gosql.Tx) error {
 			if _, err := tx.Exec(
 				"SELECT crdb_internal.unsafe_upsert_descriptor($1, decode($2, 'hex'));",
@@ -140,15 +138,12 @@ func TestDescriptorRepairOrphanedDescriptors(t *testing.T) {
 				parentID, schemaID, tableName, descID)
 			return err
 		}))
-
-		// Now that we've finished setting up the test, we can restore validations.
-		_, err := db.Exec(`SET CLUSTER SETTING sql.catalog.descs.validate_on_write.enabled = true`)
-		require.NoError(t, err)
+		descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, true)
 
 		// Ideally we should be able to query `crdb_internal.invalid_objects` but it
 		// does not do enough validation. Instead we'll just observe the issue that
 		// the parent descriptor cannot be found.
-		_, err = db.Exec(
+		_, err := db.Exec(
 			"SELECT count(*) FROM \"\".crdb_internal.tables WHERE table_id = $1",
 			descID)
 		require.Regexp(t, "internal error: desc 53: parentID 52 does not exist", err)
@@ -252,8 +247,6 @@ func TestDescriptorRepair(t *testing.T) {
 	ctx := context.Background()
 	setup := func(t *testing.T) (serverutils.TestServerInterface, *gosql.DB, func()) {
 		s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
-		_, err := db.Exec(`SET CLUSTER SETTING sql.catalog.descs.validate_on_write.enabled = false`)
-		require.NoError(t, err)
 		return s, db, func() {
 			s.Stopper().Stop(ctx)
 		}
@@ -440,12 +433,12 @@ SELECT crdb_internal.unsafe_delete_namespace_entry("parentID", 0, 'foo', id)
 			now := s.Clock().Now().GoTime()
 			defer cleanup()
 			tdb := sqlutils.MakeSQLRunner(db)
+			descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, false)
 			for _, op := range tc.before {
 				tdb.Exec(t, op)
 			}
-			_, err := db.Exec(`SET CLUSTER SETTING sql.catalog.descs.validate_on_write.enabled = true`)
-			require.NoError(t, err)
-			_, err = db.Exec(tc.op)
+			descs.ValidateOnWriteEnabled.Override(&s.ClusterSettings().SV, true)
+			_, err := db.Exec(tc.op)
 			if tc.expErrRE == "" {
 				require.NoError(t, err)
 			} else {

